src/sentry/celery.py

```diff
@@ -99,13 +103,21 @@ def delay(self, *args, **kwargs):
 
     def apply_async(self, *args, **kwargs):
         self._add_metadata(kwargs)
-        # If intended detect bad uses of pickle and make the tasks fail in tests.  This should
-        # in theory pick up a lot of bad uses without accidentally failing tasks in prod.
-        if (
+        # If there is a bad use of pickle create a sentry exception to be found and fixed later.
+        # If this is running in tests, instead raise the exception and fail outright.
+        should_complain = (
             settings.CELERY_COMPLAIN_ABOUT_BAD_USE_OF_PICKLE
             and self.name not in LEGACY_PICKLE_TASKS
-        ):
-            good_use_of_pickle_or_bad_use_of_pickle(self, args, kwargs)
+        )
+        should_sample = random.random() <= settings.CELERY_PICKLE_ERROR_REPORT_SAMPLE_RATE
+        if should_complain or should_sample:
+            try:
+                good_use_of_pickle_or_bad_use_of_pickle(self, args, kwargs)
+            except TypeError:
+                logger.exception(
+                    "Task args contain unserializable objects",
+                )
+                raise
```

Comment:
```suggestion
                if settings.CELERY_COMPLAIN_ABOUT_BAD_USE_OF_PICKLE:
                    raise
```

We don't want to raise in production, but we do want to raise in tests.